### ISSUE TITLE
Restarting the window in REPL blows up

### ISSUE THREAD

**Issue by Folcon** [HumbleUI/HumbleUI#56]
Just starting to get back into this again and it's great to see stuff has progressed a lot!

Looking at the REPL based workflow, one thing that's a bit unclear is how to restart the window now?

I'm asking this because if I change a component low down in the hierarchy such as a label within a checkbox in the app, reloading that component in the REPL doesn't change it, only restarting the REPL seems to, which is not ideal as there's a clear pause that occurs when I kill the app, about a 3rd of the time which seems to temporarily lock up the UI on occasion on my mac. I'm assuming there's some delay in letting go of the graphics card.

Trying to restart the window for example using:
```clojure
(io.github.humbleui.core/thread
  (let [screen (last (app/screens))]
    (reset! state/*window
        (ui/window
            {:title    "Humble 🐝 UI"
               :mac-icon "resources/images/icon.icns"
               :screen   (:id screen)
               :width    600
               :height   400
               :x        (if (= (:id screen) (:id (app/primary-screen))) :left :center)
               :y        :center}
            state/*app))))
```

Blows up:
```shell
2022-12-10 02:27:25.660 java[28657:9612981] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow drag regions should only be invalidated on the Main Thread!'
*** First throw call stack:
(
        0   CoreFoundation                      0x00007fff2f8d1b57 __exceptionPreprocess + 250
        1   libobjc.A.dylib                     0x00007fff685825bf objc_exception_throw + 48
        2   CoreFoundation                      0x00007fff2f8fa34c -[NSException raise] + 9
        3   AppKit                              0x00007fff2caf45ec -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 310
        4   AppKit                              0x00007fff2cadc052 -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 1416
        5   AppKit                              0x00007fff2cadbac3 -[NSWindow initWithContentRect:styleMask:backing:defer:] + 42
        6   libjwm_x64.dylib                    0x000000010bab9e77 _ZN3jwm9WindowMac4initEv + 263
        7   libjwm_x64.dylib                    0x000000010baba3e3 Java_io_github_humbleui_jwm_WindowMac__1nMake + 227
        8   ???                                 0x000000011bbdb051 0x0 + 4760383569
)
libc++abi.dylib: terminating with uncaught exception of type NSException
/bin/sh: line 1: 28657 Abort trap: 6           lein with-profile +dev trampoline run
make: *** [dev] Error 134
```

EDIT:
Ok, I found `io.github.humbleui.app/doui`, which seems to work, I'm going to have to do some overrides so that the quit on close behaviour is disabled on development.

Not sure if there's a more elegant approach?

EDIT2:
Ok, it doesn't seem to consistently work, so there's still a reloading issue, not sure if it's because the components aren't being reloaded properly or something else, I'm going to investigate using `clojure.tools.namespace.repl`.

**Comment by tonsky** [comment]
The solution I designed for REPL is that `ui/window` can not only accept component, but also a reference (e.g. a var) or an atom. I also just pushed commit 9a1f239b85d5913c2f386e8a2ef0d3f4c0760552 that even set ups watcher for you on var/atom. So just reset!/def and window should refresh! Let me know if that works

**Comment by Folcon** [comment]
Ok, so I looked over the code and I've added a `^{:clojure.tools.namespace.repl/load false}` into the core function, which should hopefully allow a full reload to occur while still having everything update.

We'll see how it works.

I've been using `force-reload` as defined:
```clojure
(ns user
  (:require [clojure.tools.namespace.repl :as ns]
            [io.github.humbleui.app :as app]
            [io.github.humbleui.window :as window]
            [example.state :as state]
            [example.core :refer [make-window]]))

(ns/set-refresh-dirs "src" "dev")

(defn reload [{:keys [hard-refresh? reflection? math?]
               :or {hard-refresh? false reflection? true math? :warn-on-boxed}}]
  (set! *warn-on-reflection* reflection?)
  (set! *unchecked-math* math?)
  (let [ns-refresh-fn
        (if hard-refresh?
          (requiring-resolve 'clojure.tools.namespace.repl/refresh-all)
          (requiring-resolve 'clojure.tools.namespace.repl/refresh))
        res (ns-refresh-fn)]
    (if (instance? Throwable res)
      (throw res)
      res)))

(defn force-reload
  ([] (force-reload {:hard-refresh? false
                     :math? nil}))
  ([opts]
   (do
     (when @state/*window
       (window/close @state/*window))
     (reload opts)
     (app/doui
       (make-window)))))

(comment
  (force-reload))
```

As re-defining a low level form doesn't always work as that doesn't change anything unless you propagate that change up the entire app tree.

----- EDIT -----

Ok, it's not quite fixed on macos, there's a new issue:
```shell
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fff32d4887d, pid=64764, tid=152339
#
# JRE version: OpenJDK Runtime Environment Homebrew (19.0.1) (build 19.0.1)
# Java VM: OpenJDK 64-Bit Server VM Homebrew (19.0.1, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-amd64)
# Problematic frame:
# C  [CoreVideo+0x187d]  CVHWTime::reset()+0xd
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/folcon/Code/pankration/hs_err_pid64764.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/Homebrew/homebrew-core/issues
#
/bin/sh: line 1: 64764 Abort trap: 6           lein with-profile +dev trampoline run
make: *** [dev] Error 134
```

I've got an error report this time, would it be useful for you to see?

**Comment by Folcon** [comment]
Ok, noticed new behavior on windows as well, using same functionality as defined above, it spawns a new window, the old window goes black and then the entire application exits with:

```clojure
Process finished with exit code 0
```

So there's no exception or anything.

Is this due to a gap / difference in the current JWM api's between macos / windows?

**Comment by tonsky** [comment]
I’m not sure. The way I had it setup is that I don’t re-create window on reload, only a var that represents “application” or “ui” in that window. So window stays the same, but it starts rendering new UI

**Comment by Folcon** [comment]
Hmm, from what I could tell I was doing a similar thing, ok, I'll try and investigate this over the weekend!

EDIT:
So I finally had a chance to properly chase this up and it's looking fixed?

I'm going to keep an eye on it for a bit longer, but if it doesn't reappear in a few weeks of fiddling, then I'll close this and try and document what the workflow looks like as well somewhere =)...

EDIT:
Spoke too soon, in some cases the UI just does not update, I've not figured out why, but when I get in this state, I can change something simple like a label and it just doesn't change.

So something is happening, interestingly the `refresh` doesn't detect a change, but neither `refresh` nor `refresh-all` in `clojure.tools.namespace.repl` fix it.

This is why I changed to reseting the window I think.

**Comment by refset** [comment]
Just adding my experience here in case it's useful - I tried setting `:exit-on-close? false` which worked to prevent the REPL from shutting down when I closed the window via my window manager, but the next time I called `ui/start-app!` (eval'ing `-main`) I got this SIGSEGV:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000728cc5c6f65d, pid=3538188, tid=3538403
#
# JRE version: OpenJDK Runtime Environment (22.0+36) (build 22+36)
# Java VM: OpenJDK 64-Bit Server VM (22+36, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# C  [libjwm_x64.so+0x1f65d]  Java_io_github_humbleui_jwm_impl_RefCounted_getRefCount+0xd
```
